code/__HELPERS/unsorted.dm

```diff
@@ -1584,4 +1584,22 @@ Game Mode config tags:
 	playsound(T, fire_sound, 50, 1)
 	A.OnFired()
 	spawn()
-		A.process()
\ No newline at end of file
+		A.process()
+
+
+//Increases delay as the server gets more overloaded,
+//as sleeps aren't cheap and sleeping only to wake up and sleep again is wasteful
+#define DELTA_CALC max(((max(world.tick_usage, world.cpu) / 100) * max(Master.sleep_delta,1)), 1)
+
+/proc/stoplag()
+	. = round(1*DELTA_CALC)
+	sleep(world.tick_lag)
+	if (world.tick_usage > TICK_LIMIT_TO_RUN) //woke up, still not enough tick, sleep for more.
+		. += round(2*DELTA_CALC)
+		sleep(world.tick_lag*2*DELTA_CALC)
+		if (world.tick_usage > TICK_LIMIT_TO_RUN) //woke up, STILL not enough tick, sleep for more.
+			. += round(4*DELTA_CALC)
+			sleep(world.tick_lag*4*DELTA_CALC)
+			//you might be thinking of adding more steps to this, or making it use a loop and a counter var
```

Comment:
Number of attempts as an argument wouldn't make sense, waking up to just sleep again causes more lag.

Adding the while loop I once advised against when i first made this proc on the other hand wouldn't be that bad of an idea. its rare but some times when some shit spawns in a loop or shit that sleeps in a loop per item (throwing, and what not) byonds sleep queue can get so big that just the sleep operation to re-insert an item into the queue can have a high cost (5-10% of a 50ms tick per tick), so decreasing how often these wake up and re-insert becomes worth it.

Most of this proc was written before stonedmc when tick_usage was new and scary.